### PR TITLE
Refactor the backend client to handle the marshaling/unmarshaling

### PR DESCRIPTION
**Pull request by robbrockbank**

### CODE REVIEW

File: lib/backend/etcd.go [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+	if results, err := c.etcdKeysAPI.Get(context.Background(), key, &etcdListOpts); err != nil {
+		return nil, err
+	} else {
+		return filterEtcdList(results.Node, l), nil
+	}
+}
+
+// Set an existing entry in the datastore.  This ignores whether an entry already
+// exists.
+func (c *EtcdClient) set(d DatastoreObject, options *etcd.SetOptions) (*DatastoreObject, error) {
```
Comment (caseydavenport) [inline lib/backend/etcd.go]: I think this should be `Set` with a capital S, yah?

Comment (robbrockbank) [inline lib/backend/etcd.go]: nah - it's just an internal helper method called from the Create/Update/Apply public methods

Comment (caseydavenport) [inline lib/backend/etcd.go]: Ah yeah,  I see.

File: lib/backend/etcd.go [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+// exists.
+func (c *EtcdClient) Apply(d *DatastoreObject) (*DatastoreObject, error) {
+	switch t := d.Key.(type) {
+	case ProfileKey:
+		return t.Apply(c, d)
+	}
+	return c.set(d, etcdApplyOpts)
+}
+
+// Delete an entry in the datastore.  This errors if the entry does not exists.
```
Comment (caseydavenport) [inline lib/backend/etcd.go]: typo: exists

File: lib/backend/etcd.go [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+		return nil, convertEtcdError(err, key)
+	} else if object, err := ParseValue(k, []byte(results.Node.Value)); err != nil {
+		return nil, err
+	} else {
+		elem := reflect.ValueOf(object).Elem().Interface()
+		return &DatastoreObject{Key: k, Object: elem, Revision: results.Node.ModifiedIndex}, nil
+	}
+}
+
+// List entries in the datastore.  This may return an empty list of there are
```
Comment (caseydavenport) [inline lib/backend/etcd.go]: typo: of -> if

**Comment by caseydavenport** [comment]
@robbrockbank This generally looks fine to me, though I've only taken a quick skim at a high level.  Are you happy to merge this and make patches on top for any lower-level issues we uncover?

**Pull request closed without merging**